src/app/plugins/platform/JavaScript/templates/webpack.config.js

```diff
@@ -28,226 +28,230 @@ const useCSSModules = false;
 const { CleanWebpackPlugin } = require('clean-webpack-plugin');
 const HtmlWebpackPlugin = require('html-webpack-plugin');
 const ExtractCssChunksPlugin = require('extract-css-chunks-webpack-plugin');
-const TerserJsPlugin = require('terser-webpack-plugin');
+const UglifyJsPlugin = require('uglifyjs-webpack-plugin');
 const ForkTsCheckerPlugin = require('fork-ts-checker-webpack-plugin');
 const OptimizeCssAssetsPlugin = require('optimize-css-assets-webpack-plugin');
 
-const isProduction = process.env.NODE_ENV === 'production';
-const isDevelopment = !isProduction;
-
-// Derived settings from variables above:
-
-const templatePrefix = isProduction ? '[id]' : '[name]';
-const templateSuffix = hashFilenames ? '-[chunkhash]' : '';
-
-const filenameTemplate = '[name]' + templateSuffix;
-const chunkFilenameTemplate = templatePrefix + templateSuffix;
-const imageFilenameTemplate = isProduction
-  ? '[hash].[ext]'
-  : '[name]-[hash].[ext]';
-
-function createFilenameTemplate(extension) {
+function createFilenameTemplate(filenameTemplate, extension) {
   return filenameTemplate + '.' + extension;
 }
 
-function createChunkFilenameTemplate(extension) {
+function createChunkFilenameTemplate(chunkFilenameTemplate, extension) {
   return chunkFilenameTemplate + '.' + extension;
 }
 
-// NB. Changing the name of the target directory requires updates to Capistrano and the
-// .gitignore file.
-const targetDirectory = path.join(__dirname, 'public');
-
-// We always use these plugins
-const plugins = [
-  // Remove stale build output
-  new CleanWebpackPlugin(),
-
-  // Create separate CSS stylesheets
-  new ExtractCssChunksPlugin({
-    filename: createFilenameTemplate('css'),
-    chunkFilename: createChunkFilenameTemplate('css'),
-  }),
-
-  // Type-check and run TSLint on all files in the project
-  new ForkTsCheckerPlugin({
-    async: false,
-    formatter: 'codeframe',
-    tslint: true,
-  }),
-
-  // Autogenerate a index.html for SPAs and webpack-dev-server.
-  new HtmlWebpackPlugin({
-    template: path.join(__dirname, 'src/index.html'),
-  }),
-];
-
 // Production plugins
 // If you're used to Webpack <= 3, you're probably wondering where we add production plugins.
 // The short answer is: we don't! Webpack 4 has built-in optimization options, and we're leaning on
 // its built-in defaults rather than overriding them.
 
-module.exports = {
-  context: __dirname,
+module.exports = env => {
+  const isProduction = env.NODE_ENV === 'production';
```

Comment:
I think I'd prefer to see this function take the `mode` parameter directly. This makes it a bit more explicit about what it's being used for, and you'll be able to avoid some work in determining the `mode` option for Webpack:

```js
module.exports = ({ mode }) => {
  const isProduction = mode === 'production';
  const isDevelopment = !isProduction;

  return {
    mode,
    // ...
  };
};

Reply:
I based that off of https://webpack.js.org/guides/environment-variables/, but I have no strong attachment to using environment variables here.

Reply:
Let's go with `mode` - it's more explicit and should help communicate to folks that they're not limited to strings when communicating with the Webpack config function.